compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +152,175 @@ pub fn expand_deriving_smart_ptr(
+                from_name: pointee_ty_ident.name,
+                to_ty: &s_ty,
+                rewritten: false,
+            };
+            substitution.visit_param_bound(&mut bound, BoundKind::Bound);
+            if substitution.rewritten {
+                params.bounds.push(bound);
+            }
+        }
+    }
 
     // Add the `__S: ?Sized` extra parameter to the impl block.
+    // We should also write the bounds from `#[pointee]` to `__S` as required by `Unsize<__S>`.
     let sized = cx.path_global(span, path!(span, core::marker::Sized));
-    let bound = GenericBound::Trait(
-        cx.poly_trait_ref(span, sized),
-        TraitBoundModifiers {
-            polarity: ast::BoundPolarity::Maybe(span),
-            constness: ast::BoundConstness::Never,
-            asyncness: ast::BoundAsyncness::Normal,
-        },
-    );
-    let extra_param = cx.typaram(span, Ident::new(sym::__S, span), vec![bound], None);
-    impl_generics.params.push(extra_param);
+    if self_bounds.iter().all(|bound| {
+        if let GenericBound::Trait(
+            trait_ref,
+            TraitBoundModifiers { polarity: ast::BoundPolarity::Maybe(_), .. },
+        ) = bound
+        {
+            !is_sized_marker(&trait_ref.trait_ref.path)
+        } else {
+            false
+        }
+    }) {
+        self_bounds.push(GenericBound::Trait(
+            cx.poly_trait_ref(span, sized),
+            TraitBoundModifiers {
+                polarity: ast::BoundPolarity::Maybe(span),
+                constness: ast::BoundConstness::Never,
+                asyncness: ast::BoundAsyncness::Normal,
+            },
+        ));
+    }
+    {
+        let mut substitution =
+            TypeSubstitution { from_name: pointee_ty_ident.name, to_ty: &s_ty, rewritten: false };
+        for bound in &mut self_bounds {
+            substitution.visit_param_bound(bound, BoundKind::Bound);
+        }
+    }
+
+    // We should also commute the where bounds from `#[pointee]` to `__S`
+    // as well as any bound that indirectly involves the `#[pointee]` type.
+    for bound in &generics.where_clause.predicates {
+        if let ast::WherePredicate::BoundPredicate(bound) = bound {
+            let bound_on_pointee = bound
+                .bounded_ty
+                .kind
+                .is_simple_path()
+                .map_or(false, |name| name == pointee_ty_ident.name);
+
+            let bounds: Vec<_> = bound
+                .bounds
+                .iter()
+                .filter(|bound| {
+                    if let GenericBound::Trait(
+                        trait_ref,
+                        TraitBoundModifiers { polarity: ast::BoundPolarity::Maybe(_), .. },
+                    ) = bound
+                    {
+                        !bound_on_pointee || !is_sized_marker(&trait_ref.trait_ref.path)
+                    } else {
+                        true
+                    }
+                })
```

Comment:
This logic deserved a detailed explanation explaining what kinds of bounds this is trying to match against.

Reply:
Specifically, I am still confused by why we have `bound_on_pointee` being special-cased for `BoundPolarity::Maybe`.

Reply:
It is possible to remove the special treatment here, if someday we can relax the restriction on repeated `?Sized` bounds on the same type.

Reply:
I suppose that I can check this earlier, so we don't need the special-casing.